File: code-go/binarytreelevelordertraversal/levelorder/level_order.go
package main

import "fmt"

// 二叉树的层序遍历
// 给你一个二叉树，请你返回其按 **层序遍历** 得到的节点值。 （即逐层地，从左到右访问所有节点）。

// 示例
// 输入: [3,9,20,null,null,15,7]
//        3
//       / \
//      9  20
//        /  \
//       15   7
// 输出: [
//        [3],
//        [9,20],
//        [15,7]
//      ]

type TreeNode struct {
    Val int
    Left *TreeNode
    Right *TreeNode
}

func levelOrder(root *TreeNode) [][]int {
    return bfs(root)

    //results := make([][]int, 0)
    //results = dfs(root, 1, results)
    //return results
}

// 广度优先
func bfs (root *TreeNode) [][]int {
    results := make([][]int, 0)
    if root == nil {
        return results
    }
    // 当做队列
    q := make([]*TreeNode, 0)
    // 根节点先入队列
    q = append(q, root)
    for len(q) > 0 {
        val := make([]int, 0)
        // 临时存放下一级节点
        p := make([]*TreeNode, 0)
        // 遍历当前层级的所有节点
        for i := 0; i < len(q); i++ {
            // 当前层级的根节点
            node := q[i]
            val = append(val, node.Val)
            if node.Left != nil {
                p = append(p, node.Left)
            }
            if node.Right != nil {
                p = append(p, node.Right)
            }
        }
        results = append(results, val)
        // 修改成下一次遍历的队列
        q = p
    }
    return results
}

// 深度优先
func dfs(node *TreeNode, level int, results [][]int) [][]int {
    if node == nil {
        return [][]int{}
    }
    if level - 1 >= len(results) {
        val := make([]int, 0)
        results = append(results, val)
    }
    // results 本身的引用会被修改
    results[level - 1] = append(results[level - 1], node.Val)

    // 递归左子节点
    leftResults := dfs(node.Left, level + 1, results)
    if len(leftResults) > 0 {
        // 重新修改结果集
        results = leftResults
    }
    // 递归右子节点
    rightResults := dfs(node.Right, level + 1, results)
    if len(rightResults) > 0 {
        results = rightResults
    }
    return results
}

func main() {
    root := &TreeNode{Val: 3}
    node1 := &TreeNode{Val: 9}
    node2 := &TreeNode{Val: 20}
    node3 := &TreeNode{Val: 15}
    node4 := &TreeNode{Val: 7}
    root.Left = node1
    root.Right = node2
    node2.Left = node3
    node2.Right = node4

    results := levelOrder(root)
    fmt.Println(results)
}

File: code-rust/src/binary_tree_level_order_traversal/level_order.rs
/// 二叉树的层序遍历
/// 给你一个二叉树，请你返回其按 **层序遍历** 得到的节点值。 （即逐层地，从左到右访问所有节点）。

/// 示例
/// 输入: [3,9,20,null,null,15,7]
///        3
///       / \
///      9  20
///        /  \
///       15   7
/// 输出: [
///        [3],
///        [9,20],
///        [15,7]
///      ]

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode { val, left: None, right: None }
    }
}

use std::rc::Rc;
use std::cell::RefCell;

pub struct Solution {}

impl Solution {
    pub fn level_order(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Vec<i32>> {
        Solution::bfs(root)

        // let mut results = Vec::new();
        // Solution::dfs(root, 1, results)
    }

    /// 广度优先
    fn bfs(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Vec<i32>> {
        let mut results = vec![];
        if root.is_none() {
            return results;
        }
        // 当做队列
        let mut q = Vec::new();
        // 根节点先入队列
        q.push(root);
        while q.len() > 0 {
            let mut val = Vec::new();
            // 临时存放下一级节点
            let mut p = Vec::new();
            // 遍历当前层级的所有节点
            for i in 0..q.len() {
                // 当前层级的根节点
                let node = q.get(i).unwrap();
                val.push(node.as_ref().unwrap().borrow().val);
                if node.as_ref().unwrap().borrow().left.is_some() {
                    p.push(node.as_ref().unwrap().borrow().left.clone());
                }
                if node.as_ref().unwrap().borrow().right.is_some() {
                    p.push(node.as_ref().unwrap().borrow().right.clone());
                }
            }
            results.push(val);
            // 修改成下一次遍历的队列
            q = p;
        }
        results
    }

    #[allow(dead_code)]
    /// 深度优先
    fn dfs(node: Option<Rc<RefCell<TreeNode>>>, level: i32, mut results: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        if node.is_none() {
            return Vec::new()
        }
        if level - 1 >= results.len() as i32 {
            results.push(Vec::new());
        }
        // results 本身的引用会被修改
        results[(level - 1) as usize].push(node.as_ref().unwrap().borrow().val);

        // 递归左子节点
        let left_results = Solution::dfs(node.as_ref().unwrap().borrow().left.clone(),
                                         level + 1, results.clone());
        if left_results.len() > 0 {
            // 重新修改结果集
            results = left_results;
        }
        // 递归右子节点
        let right_results = Solution::dfs(node.as_ref().unwrap().borrow().right.clone(),
                                         level + 1, results.clone());
        if right_results.len() > 0 {
            results = right_results;
        }
        results
    }
}

#[test]
fn test_level_order() {
    let mut root = Some(Rc::new(RefCell::new(TreeNode::new(3))));
    let node1 = Some(Rc::new(RefCell::new(TreeNode::new(9))));
    let mut node2 = Some(Rc::new(RefCell::new(TreeNode::new(20))));
    let node3 = Some(Rc::new(RefCell::new(TreeNode::new(15))));
    let node4 = Some(Rc::new(RefCell::new(TreeNode::new(7))));

    node2.as_mut().unwrap().borrow_mut().left = node3;
    node2.as_mut().unwrap().borrow_mut().right = node4;
    root.as_mut().unwrap().borrow_mut().left = node1;
    root.as_mut().unwrap().borrow_mut().right = node2;

    let results = Solution::level_order(root);
    println!("{:?}", results);
}
